File: src/exercises/stone_game_3/python/solution.py
class Solution:
    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)
        dp = [0] * 4

        for i in range(n - 1, -1, -1):
            total = 0
            dp[i % 4] = float("-inf")
            for j in range(i, min(i + 3, n)):
                total += stoneValue[j]
                dp[i % 4] = max(dp[i % 4], total - dp[(j + 1) % 4])

        if dp[0] == 0:
            return "Tie"
        return "Alice" if dp[0] > 0 else "Bob"

File: src/exercises/stone_game_3/go/solution.go
func stoneGameIII(stoneValue []int) string {
	n := len(stoneValue)
	dp := make([]int, 4)
	for i := n - 1; i >= 0; i-- {
		total := 0
		dp[i%4] = math.MinInt32
		for j := i; j < min(i+3, n); j++ {
			total += stoneValue[j]
			dp[i%4] = max(dp[i%4], total-dp[(j+1)%4])
		}
	}
	if dp[0] == 0 {
		return "Tie"
	}
	if dp[0] > 0 {
		return "Alice"
	}
	return "Bob"
}

